liblogger/src/test/java/com/mapbox/navigation/logger/MapboxLoggerTest.kt

```diff
@@ -1,21 +1,200 @@
+        val loggerObserver = mockk<LoggerObserver>(relaxed = true)
+        logger.setObserver(loggerObserver)
+        logger.logLevel = INFO
+
+        logger.i("TAG", "some message", throwable)
+
+        verify { Timber.tag("TAG") }
+        verify { Timber.i(throwable, "some message") }
+        verify { loggerObserver.log(INFO, LogEntry("TAG", "some message", throwable)) }
+    }
+
+    @Test
+    fun infoTestIncorrectLogLevel() {
+        val throwable = mockk<Throwable>()
+        val loggerObserver = mockk<LoggerObserver>(relaxed = true)
+        logger.setObserver(loggerObserver)
+        logger.logLevel = WARN
+
+        logger.i("TAG", "some message", throwable)
+
+        verify(exactly = 0) { Timber.tag(any()) }
+        verify(exactly = 0) { Timber.i(any<Throwable>(), any()) }
+        verify(exactly = 0) { loggerObserver.log(any(), any()) }
+    }
+
+    @Test
+    fun warningTestSuccessful() {
+        val throwable = mockk<Throwable>()
+        val loggerObserver = mockk<LoggerObserver>(relaxed = true)
+        logger.setObserver(loggerObserver)
+        logger.logLevel = WARN
+
+        logger.w("TAG", "some message", throwable)
+
+        verify { Timber.tag("TAG") }
+        verify { Timber.w(throwable, "some message") }
+        verify { loggerObserver.log(WARN, LogEntry("TAG", "some message", throwable)) }
+    }
+
+    @Test
+    fun warningTestIncorrectLogLevel() {
+        val throwable = mockk<Throwable>()
+        val loggerObserver = mockk<LoggerObserver>(relaxed = true)
+        logger.setObserver(loggerObserver)
+        logger.logLevel = ERROR
+
+        logger.w("TAG", "some message", throwable)
+
+        verify(exactly = 0) { Timber.tag(any()) }
+        verify(exactly = 0) { Timber.w(any<Throwable>(), any()) }
+        verify(exactly = 0) { loggerObserver.log(any(), any()) }
+    }
+
+    @Test
+    fun errorTestSuccessful() {
+        val throwable = mockk<Throwable>()
+        val loggerObserver = mockk<LoggerObserver>(relaxed = true)
+        logger.setObserver(loggerObserver)
+        logger.logLevel = ERROR
+
+        logger.e("TAG", "some message", throwable)
+
+        verify { Timber.tag("TAG") }
+        verify { Timber.e(throwable, "some message") }
+        verify { loggerObserver.log(ERROR, LogEntry("TAG", "some message", throwable)) }
+    }
+
+    @Test
+    fun errorTestIncorrectLogLevel() {
+        val throwable = mockk<Throwable>()
+        val loggerObserver = mockk<LoggerObserver>(relaxed = true)
+        logger.setObserver(loggerObserver)
+        logger.logLevel = NONE
+
+        logger.e("TAG", "some message", throwable)
+
+        verify(exactly = 0) { Timber.tag(any()) }
+        verify(exactly = 0) { Timber.e(any<Throwable>(), any()) }
+        verify(exactly = 0) { loggerObserver.log(any(), any()) }
+    }
+
+    @Test
+    fun tagSetupNotCalledWhenNullTagPassedToVerboseCall() {
+        logger.v(null, "some message", mockk())
+
+        verify(exactly = 0) { Timber.tag(any()) }
```

Comment:
Out of curiosity, what's going to be used as logcat `TAG` in this scenario?

Reply:
`2019-12-03 15:36:47.800 13587-13587/com.mapbox.services.android.navigation.app D/MapboxLogger$d: onProgressChange: fraction of route traveled: 0.028925056`
vs
`2019-12-03 15:36:47.800 13587-13587/com.mapbox.services.android.navigation.app D/TAGTAG: onProgressChange: fraction of route traveled: 0.028925056`

So as you can see, in case when we are not provide any tag -> Timber push class name and method name where it called to logcat) Actually, it was very interesting question)